server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -643,11 +643,12 @@ private static Object deepCopy(Object value) {
      * for this document.
      * @param pipeline Pipeline to execute
      * @throws Exception On exception in pipeline execution
+     * @throws IngestCycleException If an cycle (infinite loops) are found between pipelines
      */
     public IngestDocument executePipeline(Pipeline pipeline) throws Exception {
         try {
             if (this.executedPipelines.add(pipeline) == false) {
-                throw new IllegalStateException("Cycle detected for pipeline: " + pipeline.getId());
+                throw new IngestCycleException("Cycle detected for pipeline: " + pipeline.getId());
```

Comment:
This will get wrapped in a IllegalArgumentException.
```
"reason": "java.lang.IllegalArgumentException: org.elasticsearch.ingest.IngestCycleException: Cycle detected for pipeline: pipeline2"
```
Since we can only detect cycles at runtime, it is the index request that will throw this exception e.g.
```
POST test/_doc?pipeline=pipeline1
{}
```
and results in the 500 HTTP response (not the pipeline put), and thus I think 500 is appropriate since there is nothing wrong that index request. Ideally we would check for cycles for on pipeline put and throw a 400, but we can't reliable check for cycles until runtime.